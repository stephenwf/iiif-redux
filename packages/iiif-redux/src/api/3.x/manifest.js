import memoize from 'lodash.memoize';
import { createSelector } from 'reselect';
import mapById from '../../utility/mapById';
import * as technical from './iiif/technical';
import * as descriptive from './iiif/descriptive';
import * as linking from './iiif/linking';
import * as structural from './iiif/structural';
import {
  getAllAnnotationPages,
  getAllCanvases,
  getAllCanvasReferences,
  getAllContentResources,
  getAllRanges,
  getAllResources,
  getAllServices,
} from '../all';
import mapByIdOrId from '../../utility/mapByIdOrId';
import mapAllResources from '../../utility/mapAllResources';
import mapAllById from '../../utility/mapAllById';
import { standardFieldMappingFactory } from '../../../es/utility/new/standardFieldMappingFactory';

const manifest = memoize(selector => {
  /**
   * Technical properties
   *
   * - getId
   * - getType
   * - getViewingDirection
   * - getBehavior
   */
  const getId = createSelector(selector, technical.getId);

  const getType = createSelector(selector, technical.getType);

  const getViewingDirection = createSelector(
    selector,
    technical.getViewingDirection
  );

  const getBehavior = createSelector(selector, technical.getBehavior);

  /**
   * Descriptive Properties
   *
   * - getLabel
   * - getMetadata
   * - getSummary
   * - getThumbnail
   * - getPosterCanvas
   * - getRequiredStatement
   * - getRights
   * - getNavDate
   */
  const getLabel = createSelector(selector, descriptive.getLabel);

  const getMetadata = createSelector(selector, descriptive.getMetadata);

  const getSummary = createSelector(selector, descriptive.getSummary);

  const getThumbnailId = createSelector(selector, descriptive.getThumbnailId);

  const getThumbnail = createSelector(
    getThumbnailId,
    getAllContentResources,
    mapByIdOrId
  );

  const getPosterCanvasId = createSelector(
    selector,
    descriptive.getPosterCanvas
  );

  const getPosterCanvas = createSelector(
    getPosterCanvasId,
    getAllCanvases,
    mapById
  );

  const getRequiredStatement = createSelector(
    selector,
    descriptive.getRequiredStatement
  );

  const getRights = createSelector(selector, descriptive.getRights);

  const getNavDate = createSelector(selector, descriptive.getNavDate);

  /**
   * Linking properties
   *
   * - getSeeAlso
   * - getService
   * - getLogo
   * - getHomepage
   * - getRendering
   * - getPartOf
   * - getStart
   */
  const getSeeAlsoIds = createSelector(selector, linking.getSeeAlso);
  const getSeeAlso = createSelector(
    getSeeAlsoIds,
    getAllContentResources,
    mapAllById
  );

  const getServiceIds = createSelector(selector, linking.getService);
  const getService = createSelector(getServiceIds, getAllServices, mapAllById);

  const getLogoIds = createSelector(selector, linking.getLogo);
  const getLogo = createSelector(
    getLogoIds,
    getAllContentResources,
    mapAllById
  );

  const getHomepageId = createSelector(selector, linking.getHomepage);
  const getHomepage = createSelector(
    getHomepageId,
    getAllContentResources,
    mapById
  );

  const getRenderingIds = createSelector(selector, linking.getRendering);
  const getRendering = createSelector(
    getRenderingIds,
    getAllContentResources,
    mapAllById
  );

  const getPartOfId = createSelector(selector, linking.getPartOf);
  const getPartOf = createSelector(
    getPartOfId,
    getAllResources,
    mapAllResources
  );

  const getStartId = createSelector(selector, linking.getStart);
  const getStart = createSelector(
    getStartId,
    getAllCanvases,
    getAllCanvasReferences,
    (start, allCanvases, allCanvasReferences) => {
      if (!start) {
        return null;
      }
      if (start.schema === 'canvas') {
        return allCanvases[start.id];
      }
      return allCanvasReferences[start.id];
    }
  );

  /**
   * Structural Properties
   *
   * - getItems
   * - getStructures
   * - getAnnotations
   */
  const getItemIds = createSelector(selector, structural.getItems);
  const getItems = createSelector(getItemIds, getAllCanvases, mapAllById);

  const getCanvasIds = getItemIds;
  const getCanvases = getItems;

  const getStructureIds = createSelector(selector, structural.getStructures);
  const getStructures = createSelector(
    getStructureIds,
    getAllRanges,
    mapAllById
  );

  const getAnnotationIds = createSelector(selector, structural.getAnnotations);
  const getAnnotations = createSelector(
    getAnnotationIds,
    getAllAnnotationPages,
    mapAllById
  );

  return {
    getId,
    getType,
    getViewingDirection,
    getBehavior,
    getLabel,
    getMetadata,
    getSummary,
    getThumbnailId,
    getThumbnail,
    getPosterCanvasId,
    getPosterCanvas,
    getRequiredStatement,
    getRights,
    getNavDate,
    getSeeAlsoIds,
    getSeeAlso,
    getServiceIds,
    getService,
    getLogoIds,
    getLogo,
    getHomepageId,
    getHomepage,
    getRenderingIds,
    getRendering,
    getPartOfId,
    getPartOf,
    getStartId,
    getStart,
    getItemIds,
    getItems,
    getCanvasIds,
    getCanvases,
    getStructureIds,
    getStructures,
    getAnnotationIds,
    getAnnotations,
  };
});

export const mappings = standardFieldMappingFactory(manifest);

export default manifest;
