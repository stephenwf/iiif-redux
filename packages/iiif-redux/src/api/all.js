const getAllCollections = state => state.resources.collections;
const getAllExternalResources = state => state.resources.externalResources;
const getAllLayers = state => state.resources.layers;
const getAllServices = state => state.resources.services;
const getAllManifests = state => state.resources.manifests;
const getAllSequences = state => state.resources.sequences;
const getAllRanges = state => state.resources.ranges;
const getAllCanvases = state => state.resources.canvases;
const getAllCanvasReferences = state => state.resources.canvasReferences;
const getAllImages = state => state.resources.imageResources;
const getAllAnnotations = state => state.resources.annotations;
const getAllAnnotationLists = state => state.resources.annotationLists;
const getAllAnnotationPages = state => state.resources.annotationPages;
const getAllAnnotationCollections = state =>
  state.resources.annotationCollections;
const getAllContentResources = state => state.resources.contentResources;
const getAllResources = state => state.resources;

export {
  getAllCollections,
  getAllExternalResources,
  getAllLayers,
  getAllServices,
  getAllManifests,
  getAllSequences,
  getAllRanges,
  getAllCanvases,
  getAllCanvasReferences,
  getAllImages,
  getAllAnnotationLists,
  getAllAnnotationPages,
  getAllAnnotationCollections,
  getAllAnnotations,
  getAllContentResources,
  getAllResources,
};
