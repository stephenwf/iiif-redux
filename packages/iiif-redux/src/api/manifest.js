import memoize from 'lodash.memoize';
import { createSelector } from 'reselect';
import validUrl from 'valid-url';
import * as technical from './iiif-technical';
import * as descriptive from './iiif-descriptive';
import {
  getAllAnnotationLists,
  getAllExternalResources,
  getAllImages,
  getAllLayers,
  getAllRanges,
  getAllSequences,
  getAllServices,
} from './all';
import * as linking from './iiif-linking';
import * as structural from './iiif-structural';

export default memoize(selector => {
  /**************************************************
   * Technical properties
   *
   * - getId
   * - getType
   * - getViewingHint
   * - getViewingDirection
   * - getNavDate
   **************************************************/
  const getId = createSelector(selector, technical.getId);

  const getType = createSelector(selector, technical.getType);

  const getViewingHint = createSelector(
    selector,
    technical.getWhitelistedViewingHint(['individuals', 'paged', 'continuous'])
  );

  const getViewingDirection = createSelector(
    selector,
    technical.getViewingDirection
  );

  const getNavDate = createSelector(selector, technical.getNavDate);

  /**************************************************
   * Descriptive properties
   *
   * - getLabel
   * - getDescription
   * - getMetadata
   * - getAttribution
   * - getLogo
   * - getLicence
   * - getThumbnail
   **************************************************/
  const getLabel = createSelector(selector, descriptive.getLabel);

  const getDescription = createSelector(selector, descriptive.getDescription);

  const getMetadata = createSelector(selector, descriptive.getMetadata);

  const getAttribution = createSelector(selector, descriptive.getAttribution);
  const getLogo = createSelector(selector, descriptive.getLogo);

  const getLicense = createSelector(selector, descriptive.getLicense);

  const getThumbnailId = createSelector(selector, descriptive.getThumbnailId);
  const getThumbnail = createSelector(
    getThumbnailId,
    getAllImages,
    (thumbnailId, allImages) => allImages[thumbnailId] || thumbnailId
  );

  /**************************************************
   * Linking properties
   *
   * - getSeeAlso
   * - getService
   * - getRelated
   * - getRendering
   * - getWithin
   **************************************************/
  const getSeeAlsoIds = createSelector(selector, linking.getSeeAlso);
  const getSeeAlso = createSelector(
    getSeeAlsoIds,
    getAllExternalResources,
    (seeAlsoIds, allExternalResources) =>
      seeAlsoIds.map(seeAlsoId => allExternalResources[seeAlsoId])
  );

  const getServiceIds = createSelector(selector, linking.getService);
  const getService = createSelector(
    getServiceIds,
    getAllServices,
    (serviceIds, allServices) =>
      serviceIds.map(serviceId => allServices[serviceId])
  );

  const getRelatedIds = createSelector(selector, linking.getRelated);
  const getRelated = createSelector(
    getRelatedIds,
    getAllExternalResources,
    (relatedIds, allExternalResources) =>
      relatedIds.map(relatedId => allExternalResources[relatedId])
  );

  const getRenderingIds = createSelector(selector, linking.getRendering);
  const getRendering = createSelector(
    getRenderingIds,
    getAllExternalResources,
    (renderingIds, allExternalResources) =>
      renderingIds.map(renderingId => allExternalResources[renderingId])
  );

  const getWithinIds = createSelector(selector, linking.getWithin);
  const getWithin = createSelector(
    getWithinIds,
    getAllLayers,
    getAllExternalResources,
    (withinIds, allLayers, allExternalResources) =>
      withinIds
        .map(within => allLayers[within.id] || allExternalResources[within.id])
        .filter(e => e)
  );

  /**************************************************
   * Structural properties
   *
   * - getSequences (Required)
   * - getStructures / getRanges
   **************************************************/
  const getSequenceIds = createSelector(selector, structural.getSequences);
  const getSequences = createSelector(
    getSequenceIds,
    getAllSequences,
    (sequenceIds, allSequences) =>
      sequenceIds.map(sequenceId => allSequences[sequenceId])
  );

  const getRangeIds = createSelector(selector, structural.getRanges);
  const getRanges = createSelector(
    getRangeIds,
    getAllRanges,
    (rangeIds, allRanges) => rangeIds.map(rangeId => allRanges[rangeId])
  );

  const getStructureIds = getRangeIds;
  const getStructures = getRanges;

  const getOtherContentIds = createSelector(
    selector,
    structural.getOtherContent
  );
  const getOtherContent = createSelector(
    getOtherContentIds,
    getAllExternalResources,
    getAllAnnotationLists,
    (otherContentIds, allExternalResources, allAnnotationLists) =>
      otherContentIds.map(
        otherContentId =>
          allAnnotationLists[otherContentId] ||
          allExternalResources[otherContentId] || {
            '@id': otherContentId,
            label: [{ '@value': 'unknown', '@language': '@none' }],
          }
      )
  );

  return {
    selector,
    // Technical
    getId,
    getType,
    getViewingHint,
    getViewingDirection,
    getNavDate,
    // Descriptive
    getLabel,
    getDescription,
    getMetadata,
    getAttribution,
    getLogo,
    getLicense,
    getThumbnailId,
    getThumbnail,
    // Linking
    getWithinIds,
    getWithin,
    getRenderingIds,
    getRendering,
    getRelatedIds,
    getRelated,
    getServiceIds,
    getService,
    getSeeAlsoIds,
    getSeeAlso,
    // Structural
    getSequenceIds,
    getSequences,
    getRangeIds,
    getRanges,
    getStructureIds,
    getStructures,
    getOtherContentIds,
    getOtherContent,
  };
});
