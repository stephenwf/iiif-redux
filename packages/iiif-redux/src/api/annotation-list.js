import { getAllAnnotationLists } from './all';
import byIdSelectorFactory from '../utility/byIdSelectorFactory';
import resourceListSelectorFactory from '../utility/resourceListSelectorFactory';
import presentation2annotationList from './2.x/annotation-list';

const annotationList = presentation2annotationList;

export default annotationList;

export const annotationLists = resourceListSelectorFactory(
  getAllAnnotationLists,
  annotationList
);

export const annotationListByIdSelector = byIdSelectorFactory(
  annotationList,
  'annotationLists'
);
