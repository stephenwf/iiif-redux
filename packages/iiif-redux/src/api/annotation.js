import byIdSelectorFactory from '../utility/byIdSelectorFactory';
import resourceListSelectorFactory from '../utility/resourceListSelectorFactory';
import presentation2Annotation from './2.x/annotation';
import { getAllAnnotations } from './all';

const annotation = presentation2Annotation;

export default annotation;

export const annotations = resourceListSelectorFactory(
  getAllAnnotations,
  annotation
);

export const annotationByIdSelector = byIdSelectorFactory(
  annotation,
  'annotations'
);
