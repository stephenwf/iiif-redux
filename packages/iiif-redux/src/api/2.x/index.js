import annotationList, {
  mappings as annotationListMapping,
} from './annotation-list';
import annotation, { mappings as annotationMapping } from './annotation';
import canvas, { mappings as canvasMapping } from './canvas';
import collection, { mappings as collectionMapping } from './collection';
import externalResource, {
  mappings as externalResourceMapping,
} from './external-resource';
import imageResource, {
  mappings as imageResourceMapping,
} from './image-resource';
import manifest, { mappings as manifestMapping } from './manifest';
import range, { mappings as rangeMapping } from './range';
import sequence, { mappings as sequenceMapping } from './sequence';

export const mappings = {
  annotationList: annotationListMapping,
  annotation: annotationMapping,
  canvas: canvasMapping,
  collection: collectionMapping,
  externalResource: externalResourceMapping,
  imageResource: imageResourceMapping,
  manifest: manifestMapping,
  range: rangeMapping,
  sequence: sequenceMapping,

  // Aliases.
  annotationPage: annotationListMapping,
  contentResources: imageResourceMapping, // @todo change when image/external get merged.
};

export {
  annotationCollection,
  annotationList,
  annotation,
  canvas,
  collection,
  externalResource,
  imageResource,
  manifest,
  range,
  sequence,
};
