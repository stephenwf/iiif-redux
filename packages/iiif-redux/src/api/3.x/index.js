import annotationCollection, {
  mappings as annotationCollectionMapping,
} from './annotation-collection';
import annotationPage, {
  mappings as annotationPageMapping,
} from './annotation-page';
import annotation, { mappings as annotationMapping } from './annotation';
import canvas, { mappings as canvasMapping } from './canvas';
import collection, { mappings as collectionMapping } from './collection';
import contentResource, {
  mappings as contentResourceMapping,
} from './content-resource';
import manifest, { mappings as manifestMapping } from './manifest';
import range, { mappings as rangeMapping } from './range';

export const mappings = {
  annotationCollection: annotationCollectionMapping,
  annotationPage: annotationPageMapping,
  annotation: annotationMapping,
  canvas: canvasMapping,
  collection: collectionMapping,
  contentResource: contentResourceMapping,
  manifest: manifestMapping,
  range: rangeMapping,
};

export {
  annotationCollection,
  annotationPage,
  annotation,
  canvas,
  collection,
  contentResource,
  manifest,
  range,
};
