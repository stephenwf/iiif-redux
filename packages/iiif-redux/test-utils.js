import createStore from './src/createStore';
import {
  selectCanvas,
  selectManifest,
  selectSequence,
} from './src/spaces/routing';
import * as currentManifest from './src/api/current-manifest';
import * as currentSequence from './src/api/current-sequence';
import * as currentCanvas from './src/api/current-canvas';
import * as reducers from './src/reducers';
import { combineReducers } from 'redux';

export function waitForRequest(store, id) {
  return new Promise(resolve => {
    store.subscribe(() => {
      const newState = store.getState();
      if (
        newState.dereferenced &&
        newState.dereferenced[id] &&
        newState.dereferenced[id].loading === false
      ) {
        resolve();
      }
    });
  });
}

const reducer = combineReducers(reducers);
export function makeStateFromActions(...actions) {
  return [{ type: '@@redux/INIT' }, ...actions].reduce(reducer, {});
}

export async function createStoreAndImportManifest(manifestJson) {
  const id = manifestJson['@id'];
  const store = createStore();
  fetch.mockResponseOnce(JSON.stringify(manifestJson));

  const whenRequestFinishes = waitForRequest(store, id);

  store.dispatch(selectManifest({ id }));

  await whenRequestFinishes;

  const sequences = currentManifest.getSequenceIds(store.getState());
  store.dispatch(selectSequence({ id: sequences[0] }));

  const canvases = currentSequence.getCanvasIds(store.getState());
  store.dispatch(selectCanvas({ id: canvases[0] }));

  return store;
}
