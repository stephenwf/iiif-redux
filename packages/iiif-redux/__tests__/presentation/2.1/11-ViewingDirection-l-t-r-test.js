import { createStructuredSelector } from 'reselect';
import testJson from '../../fixtures/presentation/2.1/12/manifest';
import { createStoreAndImportManifest } from '../../../test-utils';
import { manifestByIdSelector } from '../../../src/api/manifest';

describe('iiif/presentation/2.1/12 ViewingDirection l-t-r', () => {
  global.fetch = require('jest-fetch-mock');

  it('should render correct viewing direction', async () => {
    const store = await createStoreAndImportManifest(testJson);
    const state = store.getState();

    expect(
      manifestByIdSelector(
        currentManifest => ({
          service: currentManifest.getViewingDirection,
        }),
        {
          getId: () => testJson['@id'],
        }
      )(state)
    ).toMatchSnapshot();
  });
});
