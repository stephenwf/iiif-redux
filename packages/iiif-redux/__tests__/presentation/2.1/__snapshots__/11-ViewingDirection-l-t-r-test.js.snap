// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iiif/presentation/2.1/12 ViewingDirection l-t-r should render correct viewing direction 1`] = `
Object {
  "service": "right-to-left",
}
`;
